src/mscorlib/shared/System/Span.NonGeneric.cs

```diff
@@ -341,6 +341,44 @@ internal static unsafe bool EndsWithOrdinalHelper(ReadOnlySpan<char> span, ReadO
             return OrdinalHelper(span.Slice(span.Length - value.Length), value, value.Length);
         }
 
+        /// <summary>
+        /// Helper method for MemoryExtensions.AsSpan(T[] array, int start).
+        /// </summary>
+        public static Span<T> AsSpan<T>(T[] array, int start)
+        {
+            if (array == null)
+            {
+                if (start != 0)
+                    ThrowHelper.ThrowArgumentOutOfRangeException();
+                return default;
+            }
+            if (default(T) == null && array.GetType() != typeof(T[]))
```

Comment:
We don't do this check if we construct a ReadOnlySpan, but now that we plan to remove AsReadOnlySpan overloads, using AsSpan (and casting) has different semantics/perf than creating ReadOnlySpan using the ctor. Is this OK?

cc @jkotas, @KrzysztofCwalina

Reply:
Well, if you want to write fast code, you kind of want to use the constructor directly anyway. The various `AsXXX` methods will always tend to be slower than direct constructor calls.

Reply:
> This just makes the difference more profound.

I see. Aside from perf, if you have a covariant array and you want to create a ReadOnlySpan, you have to use the ctor now since AsSpan would throw.

Reply:
I think it is unfortunate that we are removing the AsReadOnlySpan extension methods for arrays. They existed for a reason.

Reply:
My mental model is that people would use AsSpan for most scenarios. Yes, AsSpan has limitations/cost. For those that don't like the limitations/cost, they would use the ctor. The ctor is not as convinient (not inference), but it is fast and lets users opt out from covariance checks.
Is that not a reasonable story?

Reply:
> but it is fast

Did we end up keeping all the ctors, or if you want to use a ctor do you end up having to use the longest overload that then incurs range checks you might have otherwise avoided?